cmd/prometheus-aws-discovery/main.go

```diff
@@ -23,66 +22,103 @@ func init() {
+	flag.StringVar(&tag, "tag", "", "Tag-Key used to look for exporter data (V2-Tag)")
+	flag.StringVar(&outputType, "output", "kubernetes", "Allowed Values {kubernetes|file}")
+	flag.StringVar(&filePath, "file-path", "", "Target file path for sd_config file output")
+	flag.StringVar(&kubeconfig, "kube-config", "", "Path to a kubeconfig file")
+	flag.StringVar(&namespace, "kube-namespace", "", "Namespace where to create or update configmap (If in cluster and no namespace is provided it tries to detect namespace from incluster config otherwise it uses 'default' namespace)")
+	flag.StringVar(&configmapName, "kube-configmap-name", "", "Name of the configmap to create or update with discovery output")
+	flag.StringVar(&configmapKey, "kube-configmap-key", "", "Name of configmap key to set discovery output to")
+	flag.StringVar(&iaasCSV, "iaas", "aws", "CSV of Clouds to check [aws/azure] e.g. aws,azure ")
+	flag.StringVar(&subscrID, "azure-subscr", "", "Azure Subscription ID to look for VMs")
 	verbose := flag.Bool("verbose", false, "Print verbose log messages")
 	printVersion := flag.Bool("version", false, "Print version")
 	flag.Parse()
 
+	flag.Visit(func(f *flag.Flag) {
+		if f.Name == "tagprefix" {
+			tagPrefix = true
+		}
+	})
 	if *verbose {
 		log.SetLevel(log.DebugLevel)
 	} else {
 		log.SetLevel(log.InfoLevel)
 	}
-
 	if *printVersion {
 		log.Info("Version: " + Version)
 		os.Exit(0)
 	}
 
-	validateArg("outputtype", *outputType, []string{"kubernetes", "file"})
+	log.Info("Infra " + iaasCSV)
+	infraReader := csv.NewReader(strings.NewReader(iaasCSV))
+	records, err := infraReader.ReadAll()
 
-	awsSession := session.New()
-	awsConfig := &aws.Config{}
-	ec2Client := ec2.New(awsSession, awsConfig)
-
-	log.Info("Start discovery of ec2 instances")
-	d := discovery.NewDiscovery(ec2Client, *tagPrefix)
-	instances, err := d.Discover()
 	if err != nil {
-		panic(err)
+		log.Fatal(err)
 	}
-
 	var o output.Output
-	if *outputType == "kubernetes" {
+	switch outputType {
+	case "stdout":
+		log.Info("Configured stdout as output target")
+		o = output.OutputStdOut{}
+	case "file":
+		log.Info("Configured file as output target")
+		o = output.OutputFile{FilePath: filePath}
+	default:
 		log.Info("Configured kubernetes as output target")
-		k8s, err := outputkubernetes.NewOutputKubernetes(*kubeconfig, *namespace, *configmapName, *configmapKey)
+		k8s, err := output.NewOutputKubernetes(kubeconfig, namespace, configmapName, configmapKey)
 		if err != nil {
 			panic(err)
 		}
-
 		o = k8s
-	} else {
-		log.Info("Configured file as output target")
-		o = outputfile.OutputFile{FilePath: *filePath}
 	}
-	e := o.Write(*instances)
-	if e != nil {
-		panic(e)
+
+	for _, runInfra := range records[0] {
+		log.Info(runInfra)
+		clients := []discovery.DiscoveryClient{}
+		switch runInfra {
+		case "aws":
+			log.Info("starting aws discovery")
+			validateArg("outputtype", outputType, []string{"kubernetes", "file", "stdout"})
```

Comment:
Shouldn't this validation happen before the provider based selection?
It seems only valid for aws here?

Reply:
done